lib/rules/require-camelcase-or-uppercase-identifiers.js

```diff
@@ -115,9 +190,78 @@ module.exports.prototype = {
     check: function(file, errors) {
         file.iterateTokensByType('Identifier', function(token) {
             var value = token.value;
-            if (value.replace(/^_+|_+$/g, '').indexOf('_') === -1 || value.toUpperCase() === value) {
+            /* Leading and trailing underscores signify
+             * visibility/scope and do not affect validation of the
+             * rule.  Remove them to simplify the checks. */
+            var match = /^_+/.exec(value);
+            var isPrivate = false;
+            if (match) {
+                isPrivate = true;
+                value = value.substr(match[0].length);
+            }
+            match = /_+$/.exec(value);
+            if (match) {
+                value = value.substr(0, match.index);
+            }
+
+            /* Strip at most one prefix permitted text and underscore
+             * from the identifier.  This transformation cannot change
+             * an acceptable identifier into an unacceptable
+             * identifier so we can continue with the normal
+             * verification of whatever it produces. */
+            var i;
+            var len;
+            if (this._allowedPrefixes) {
+                for (i = 0, len = this._allowedPrefixes.length; i < len; ++i) {
+                    var prefix = this._allowedPrefixes[i];
+                    if (typeof prefix === 'string') {
+                        if (((1 + prefix.length) < value.length) &&
```

Comment:
You gotta decompose here, three included `if`'s with pretty complicated conditions, whereas two of them are included in `for`.

Cyclomatic complexity should be much, much lower.